eras/babbage/impl/src/Cardano/Ledger/Babbage/TxBody.hs

```diff
@@ -0,0 +1,805 @@
+    | Just Refl <- sameNat (Proxy @(SizeHash (CC.ADDRHASH (Crypto era)))) (Proxy @28) ->
+      (compactAddr (decodeAddress28 stakeRef a b c d), toCompactValue adaVal, SNothing)
+  TxOut_AddrHash28_AdaOnly_DataHash32 stakeRef a b c d adaVal e f g h
+    | Just Refl <- sameNat (Proxy @(SizeHash (CC.HASH (Crypto era)))) (Proxy @32),
+      Just Refl <- sameNat (Proxy @(SizeHash (CC.ADDRHASH (Crypto era)))) (Proxy @28) ->
+      ( compactAddr (decodeAddress28 stakeRef a b c d),
+        toCompactValue adaVal,
+        SJust (decodeDataHash32 e f g h)
+      )
+  TxOut_AddrHash28_AdaOnly {} -> error "Impossible: Compacted and address or hash of non-standard size"
+  TxOut_AddrHash28_AdaOnly_DataHash32 {} -> error "Impossible: Compacted and address or hash of non-standard size"
+  where
+    toCompactValue :: CompactForm Coin -> CompactForm (Core.Value era)
+    toCompactValue ada =
+      fromMaybe (error "Failed to compact a `Coin` as `CompactForm (Core.Value era)`")
+        . toCompact
+        . inject
+        $ fromCompact ada
+
+viewTxOut ::
+  forall era.
+  Era era =>
+  TxOut era ->
+  (Addr (Crypto era), Core.Value era, StrictMaybe (DataHash (Crypto era)))
+viewTxOut (TxOutCompact bs c) = (addr, val, SNothing)
+  where
+    addr = decompactAddr bs
+    val = fromCompact c
+viewTxOut (TxOutCompactDH bs c dh) = (addr, val, SJust dh)
+  where
+    addr = decompactAddr bs
+    val = fromCompact c
+viewTxOut (TxOutCompactDatum bs c datum) = (addr, val, SJust dh)
+  where
+    addr = decompactAddr bs
+    val = fromCompact c
+    dh = hashData datum
+viewTxOut (TxOut_AddrHash28_AdaOnly stakeRef a b c d adaVal)
+  | Just Refl <- sameNat (Proxy @(SizeHash (CC.ADDRHASH (Crypto era)))) (Proxy @28) =
+    let addr = decodeAddress28 stakeRef a b c d
+     in (addr, inject (fromCompact adaVal), SNothing)
+viewTxOut (TxOut_AddrHash28_AdaOnly_DataHash32 stakeRef a b c d adaVal e f g h)
+  | Just Refl <- sameNat (Proxy @(SizeHash (CC.HASH (Crypto era)))) (Proxy @32),
+    Just Refl <- sameNat (Proxy @(SizeHash (CC.ADDRHASH (Crypto era)))) (Proxy @28) =
+    let addr = decodeAddress28 stakeRef a b c d
+     in (addr, inject (fromCompact adaVal), SJust (decodeDataHash32 e f g h))
+viewTxOut (TxOut_AddrHash28_AdaOnly {}) = error "Impossible: Compacted and address or hash of non-standard size"
+viewTxOut (TxOut_AddrHash28_AdaOnly_DataHash32 {}) = error "Impossible: Compacted and address or hash of non-standard size"
+
+instance
+  ( Era era,
+    Show (Core.Value era),
+    Show (CompactForm (Core.Value era))
+  ) =>
+  Show (TxOut era)
+  where
+  show = show . viewTxOut
+
+deriving via InspectHeapNamed "TxOut" (TxOut era) instance NoThunks (TxOut era)
+
+pattern TxOut ::
+  forall era.
+  ( Era era,
+    Compactible (Core.Value era),
+    Val (Core.Value era),
+    HasCallStack
+  ) =>
+  Addr (Crypto era) ->
+  Core.Value era ->
+  StrictMaybe (DataHash (Crypto era)) ->
+  TxOut era
+pattern TxOut addr vl dh <-
```

Comment:
This doesn't make sense. How anyone would ever construct a TxOut with `Data`? If I had to guess this should be something like this instead:
```suggestion
  Addr (Crypto era) ->
  Core.Value era ->
  Either (DataHash (Crypto era)) (Data era) ->
  TxOut era
pattern TxOut addr vl ed <-
```

Reply:
I don't think either works either, since there are three cases: no datum, datum hash, or inline datum. Maybe we use `Maybe Either`?

Reply:
Makes sense. I think a custom data type would be more appropriate:

```haskell
data Datum era
  = NoDatum 
  | DatumHash !(DataHash (Crypto era))
  | Datum !(Data era)
```